smithy-utils/src/main/java/software/amazon/smithy/utils/IoUtils.java

```diff
@@ -227,46 +228,90 @@ public static int runCommand(String command, Path directory, Appendable output,
+     * @param directory Directory to use as the working directory.
+     * @param input     Data to write to the stdin of the process. Use {@code null} to send no input.
+     * @param output    Sink destination for lines from stdout and stderr of the process.
+     * @param env       Environment variables to set, possibly overriding specific inherited variables.
+     * @return          Returns the exit code of the process.
+     */
+    public static int runCommand(
+            List<String> args,
+            Path directory,
+            InputStream input,
+            Appendable output,
+            Map<String, String> env
+    ) {
         ProcessBuilder processBuilder = new ProcessBuilder(args)
-                .directory(directory.toFile())
+                .directory(Objects.requireNonNull(directory.toFile(), "Process directory cannot be null"))
                 .redirectErrorStream(true);
-
-        if (!env.isEmpty()) {
-            processBuilder.environment().putAll(env);
-        }
+        processBuilder.environment().putAll(env);
 
         try {
             Process process = processBuilder.start();
-            try (BufferedReader bufferedReader = new BufferedReader(
-                    new InputStreamReader(process.getInputStream(), Charset.defaultCharset()))) {
+
+            if (input != null) {
+                OutputStream outputStream = process.getOutputStream();
+                copyInputToOutput(input, outputStream);
+                closeAndFlushStream(args, outputStream);
+            }
+
+            try (BufferedReader bufferedStdoutReader = new BufferedReader(
+                    new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
                 String line;
-                while ((line = bufferedReader.readLine()) != null) {
+                while ((line = bufferedStdoutReader.readLine()) != null) {
                     output.append(line).append(System.lineSeparator());
                 }
             }
-            process.waitFor();
-            process.destroy();
-            return process.exitValue();
+
+            return process.waitFor();
         } catch (InterruptedException | IOException e) {
             throw new RuntimeException(e);
         }
     }
 
+    // Can be replaced with InputStream#transferTo in Java 9+.
+    private static void copyInputToOutput(InputStream source, OutputStream sink) throws IOException {
+        byte[] buffer = new byte[8192];
+        int read;
+        while ((read = source.read(buffer, 0, 8192)) >= 0) {
+            sink.write(buffer, 0, read);
+        }
+    }
+
+    private static void closeAndFlushStream(List<String> args, OutputStream outputStream) {
+        try {
+            outputStream.flush();
+            outputStream.close();
+        } catch (IOException e) {
+            // TODO: Why is this necessary?
+            LOGGER.log(Level.WARNING, "Unable to close and flush outputStream for " + args, e);
```

Comment:
I'd say, if you decide to keep the flush, "flush and close" instead, as someone might think, oh, it's because you cannot flush after close :). But other than that I'm not sure why there's a `TODO` there, I guess it's necessary because any IO can throw, right?